### ISSUE TITLE
Plutus predicate failure to use Plutus error instead of string

### ISSUE THREAD

**Issue by JaredCorduan** [IntersectMBO/cardano-ledger#2513]
It would be helpful to downstream libraries if the plutus script predicate returned the actual plutus error instead of a string:

https://github.com/input-output-hk/cardano-ledger-specs/blob/8ef01f28e31789503879f1391b230aa1e50c6ab2/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L632

**Comment by MaximilianAlgehed** [comment]
I agree that this is a problem. We've bumped into issues / warts / frustrations related to this when implementing testing tools for Plutus contracts. For example over at https://github.com/input-output-hk/plutus/pull/4023.

As for solutions I don't know what's best but the information that's needed by downstream code is at least:
1. What term caused the error (in the case of e.g. builtins failing)
2. What error message did the contract developer intend to throw. Right now this information is propagated via the log but in a perfect world the interpreter would put this error in the datastructure it returns when a computation fails.

**Comment by JaredCorduan** [comment]
We unfortunately cannot run Plutus scripts in verbose mode inside the node, but there will soon be two ways to see the verbose logs.

* The CLI will soon be using the newer version of [evaluateTransactionExecutionUnits](https://github.com/input-output-hk/cardano-ledger-specs/blob/3f2c3fc6d8fc89e49cef06e514d9eef8ad737a13/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tools.hs#L103) when building transactions, so that when a Plutus scripts fails it can return the logs.
* When the node returns the [predicate failure for a failed script](https://github.com/input-output-hk/cardano-ledger-specs/blob/3f2c3fc6d8fc89e49cef06e514d9eef8ad737a13/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L420), in addition to the `show e` that is inside the `Text`, there is debug information in the `ByteString`. I believe our plan is to make using the debug information to retrieve the verbose logs available from the CLI. This is just a matter of running [debugPutus](https://github.com/input-output-hk/cardano-ledger-specs/blob/3f2c3fc6d8fc89e49cef06e514d9eef8ad737a13/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L486) with the information from the predicate failure.

**Comment by lehins** [comment]
I think we can now close this issue.
When plutus script fails upon transaction validation then the failure is reported and the whole [`PlutusWithContext`](https://github.com/IntersectMBO/cardano-ledger/blob/0a16648436f1fa5d655e5ca471d48257e3a71aec/libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/Evaluate.hs#L97-L120) is hex encoded, which later can be used in isolation with the help of [`plutus-debug`](https://github.com/IntersectMBO/cardano-ledger/blob/0a16648436f1fa5d655e5ca471d48257e3a71aec/libs/cardano-ledger-core/cardano-ledger-core.cabal#L239) tool to reproduce the error with logs (i.e. running in verbose mode) and ability to override various aspects of execution: script itself, execution units, protocol version, cost model, etc.

FTR. We do not want to run plutus in verbose mode when validating transactions, so this IMHO is the best that we can do.
